[BEAM-2072,BEAM-2092] Fixed MicrobatchSource.reader stops reading unexpectedly.

staslev opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

staslev commented:
Run Spark ValidatesRunner

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -201,56 +201,66 @@ public int hashCode() {
    */
```
This looks like either a rewrite or a near-clone of `BoundedReadFromUnboundedSource` -- any opportunity for code reuse?

staslev replied:
They definitely look clone-ish. A brief look indicates that the internal stuff in `BoundedReadFromUnboundedSource` are quite private so extending them in order perform the Spark-specific tweaks may require exposing these internals at the very least. 
Perhaps that was the reason it was cloned in the first place.

I'll try to take a deeper look and see what it would take to consolidate them.

staslev replied:
@dhalperi , while I was not able to fully remove the duplicate code, I did some cleanups that simplify things a bit.

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -201,56 +201,66 @@ public int hashCode() {
+          maxReadTime,
+          maxNumRecords);
+
+      prepareForNewBatchReading();
```
Fascinating. If I understand BEAM-2072, the issue is that you're re-starting a BoundedReader after stopping it? That's definitely not an expected use case. We expect that a `BoundedReader` is only read once, and that the way you "keep reading from that source" is by splitting off the remainder of the source and creating a new reader from the remainder.

Hmm...

Conversely to my prior comment about sharing with BoundedReadFromUnboundedSource, maybe it's worth dropping the BoundedReader-ness altogether and trying to build a simpler thing that integrates more tightly with runner API and meets the way you want to read from the source.

staslev replied:
@dhalperi , thanks for taking a look.

As of [BEAM-1294](https://issues.apache.org/jira/browse/BEAM-1294), readers are cached and thus maintain their state across multiple (Spark) batches. In order to have the `recordsCount` count elements per-batch, as opposed counting them per the reader's entire life-cycle, it needs to be re-set to `0`.

 At the moment the best place to perform this reset seems to be the `Reader#start()` method, however, as you (and the javadoc) point out, the `Reader#start()` method was originally designed to be called only.

I believe the motivation behind [BEAM-1294](https://issues.apache.org/jira/browse/BEAM-1294) was to avoid creating new readers for each new batch, as potentially there could be sources for which this would be a costly.

Are you aware of other sources / scenarios where readers are cached in a similar manner?

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -201,56 +201,66 @@ public int hashCode() {
     }
 
+    private boolean startIfNeeded() throws IOException {
+      return !started && ((started = true) && unboundedReader.start());
```
looks like `recordsRead` will be `0L` even if the reader is successfully started. Is that an Obiwan error? Maybe add some unit tests to catch BEAM-2072 and validate element counts?

staslev replied:
Consider the following snippet [taken from here](https://github.com/apache/beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/StateSpecFunctions.java#L168-L173):

```java
          boolean finished = !microbatchReader.start();
          while (!finished) {
            final WindowedValue<T> wv =
                WindowedValue.of((T) microbatchReader.getCurrent(),
                                 microbatchReader.getCurrentTimestamp(),
                                 GlobalWindow.INSTANCE,
                                 PaneInfo.NO_FIRING);
            readValues.add(CoderHelpers.toByteArray(wv, coder));
            finished = !microbatchReader.advance();
          }
```

`recordsRead` lags by 1, and essentially counts the records that have already been added to `readValues`. 

After `start()` is called for the first time, it advances the reader to the first record, and the corresponding `recordsRead` update takes place in `microbatchReader.advance()` at the end of the `while` loop.

dhalperi replied:
+1

staslev commented:
Run Spark ValidatesRunner

staslev commented:
retest this please.

staslev commented:
retest this please

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -70,13 +64,13 @@
+      final long maxNumRecords,
+      final int splitId,
+      final int sourceId,
+      final double readerCacheInterval) {
```
these `final` don't do anything

staslev replied:
Yeah, having experienced the joy of Scala's `val`, I use the make-every-thing-you-can-final inspection in IntelliJ.

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -86,12 +80,23 @@
 
+  private static synchronized void initReaderCache(final long readerCacheInterval) {
+    if (readerCache == null) {
+      LOG.info("Creating reader cache. Cache interval = " + readerCacheInterval + " ms.");
```
could use logging format `{}` instead of string concatenation, to defer string concatenation unless logging is enabled.

staslev replied:
Good point, missed it.

dhalperi commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -86,12 +80,23 @@
-  private static long[] splitNumRecords(long numRecords, int numSplits) {
-    long[] splitNumRecords = new long[numSplits];
+  private static long[] splitNumRecords(final long numRecords, final int numSplits) {
+    final long[] splitNumRecords = new long[numSplits];
```
these `final` don't do anything.

You typically want `final` for member variables that should not change after construction, and for variables you're going to pass into anonymous inner classes.

staslev replied:
I prefer all the variables to be `final` by default.

staslev commented:
retest this please.

staslev commented:
Run Spark ValidatesRunner

staslev commented:
@aviemzur

staslev commented on runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java:
```diff
@@ -193,64 +175,79 @@ public int hashCode() {
-        return false;
+        advanced = false;
+      } else {
+        advanced = advanceWithBackoff();
```
It's 3 LOC actually, since the definition of `advanced` can then also be removed.
While I usually prefer the more readable version even if it's longer, I tend to agree that we can make (throw?) an expectation here.

staslev commented:
Addressed the comments.

The [answer](https://github.com/apache/beam/pull/2733#issuecomment-298225408) to this questions is also the same as in #2733.

staslev commented:
Rebased against master.

staslev commented:
Run Spark ValidatesRunner
